ui/components/institutional/qr-code-modal/qr-code-modal.js

```diff
@@ -0,0 +1,254 @@
+import {
+  TextAlign,
+  TextColor,
+  TextVariant,
+  FontWeight,
+} from '../../../helpers/constants/design-system';
+import { I18nContext } from '../../../contexts/i18n';
+import Spinner from '../../ui/spinner';
+import {
+  getChannelId,
+  getConnectionRequest,
+} from '../../../ducks/institutional/institutional';
+import { mmiActionsFactory } from '../../../store/institutional/institution-background';
+
+export default function QRCodeModal({
+  onClose,
+  custodianName = 'custodian',
+  custodianURL,
+  setQrConnectionRequest,
+}) {
+  const t = useContext(I18nContext);
+  const dispatch = useDispatch();
+  const channelId = useSelector(getChannelId);
+  const connectionRequest = useSelector(getConnectionRequest);
+  const [publicKey, setPublicKey] = useState(null);
+  const [privateKey, setPrivateKey] = useState(null);
+  const [decryptedMessage, setDecryptedMessage] = useState(null);
+  const [qrCodeValue, setQrCodeValue] = useState(null);
+  const [error, setError] = useState('');
+  const [traceId] = useState(uuid());
+  const mmiActions = mmiActionsFactory();
+
+  const handleClose = useCallback(async () => {
+    await dispatch(mmiActions.setConnectionRequest(null));
+    onClose();
+  }, [dispatch, mmiActions, onClose]);
+
+  const handleError = useCallback((message, e) => {
+    const errorMessage = `${message} Please try again or contact support if the problem persists.`;
+    console.error(message, e);
+    setError(errorMessage);
+    captureException(e);
+  }, []);
+
+  const generateKeyPair = useCallback(async () => {
+    try {
+      const keyPair = await window.crypto.subtle.generateKey(
+        {
+          name: 'RSA-OAEP',
+          modulusLength: 4096,
+          publicExponent: new Uint8Array([1, 0, 1]),
+          hash: { name: 'SHA-256' },
+        },
+        true,
+        ['encrypt', 'decrypt'],
+      );
+
+      const exportedPublicKey = await window.crypto.subtle.exportKey(
+        'spki',
+        keyPair.publicKey,
+      );
+
+      setPublicKey(
+        btoa(String.fromCharCode(...new Uint8Array(exportedPublicKey))),
+      );
+
+      setPrivateKey(keyPair.privateKey);
+    } catch (e) {
+      handleError('An error occurred while generating cryptographic keys.', e);
+    }
+  }, [handleError]);
+
+  const decrypt = useCallback(
+    async (_privateKey, encryptedData) => {
+      try {
+        const encryptedBuffer = Uint8Array.from(atob(encryptedData), (c) =>
+          c.charCodeAt(0),
+        );
+        const decrypted = await window.crypto.subtle.decrypt(
+          { name: 'RSA-OAEP' },
+          _privateKey,
+          encryptedBuffer,
+        );
+        const decryptedString = new TextDecoder().decode(decrypted);
+        return JSON.parse(decryptedString);
+      } catch (e) {
+        handleError('An error occurred while decrypting data.', e);
+        throw e;
+      }
+    },
+    [handleError],
+  );
+
+  useEffect(() => {
+    generateKeyPair();
```

Comment:
question: do we need to inject more dependency here so the request of sending keys are not fired whenever component has any re-render?

Reply:
I remove the dependency, and now will only run on the component mount